### DISCUSSION TITLE
How to add Recurring task dynamically?

### DISCUSSION THREAD

**Discussion by 0x01F4** [kagkarlsson/db-scheduler#550] [Q&A] [answered]
I don't have the task definition before starting the scheduler. If I schedule a task after start then it is not part of known task so it is skipped. I see only option to override the TaskResolver

Also what is difference with task_name and task_id?
Can we have same task name with different task_id?

**Comment by beilCrxmarkets** [comment, marked as answer]
I think you are looking for `RecurringTaskWithPersistentSchedule`.

### ✅ How to Add a Dynamic Recurring Task
1. **Define your task** using:
   ```java
   Tasks.recurringWithPersistentSchedule(...)
   ```
2. **Create a schedulable instance**:
   ```java
   RecurringTaskWithPersistentSchedule.schedulableInstance(String id, T data)
   ```
3. **Schedule it manually** via:
   ```java
   SchedulerClient.scheduleIfNotExists(SchedulableInstance<T> schedulableInstance)
   ```

---

### 🔍 Key Differences Between Task Types
- **`RecurringTask`**
  - Static recurring task with a **fixed schedule**.
  - **Only one possible instance**.
  - Automatically scheduled at application startup.

- **`RecurringTaskWithPersistentSchedule`**
  - Dynamic recurring task with a **modifiable schedule**.
  - **Multiple possible instances**.
  - Must be scheduled **manually** via `SchedulerClient`.

---

### 🆔 Understanding `task_name` vs `task_id`
- A `TaskInstance` identifies a specific instance of a task and consists of:
  - **`task_name`** → Bound to the task definition.
  - **`task_id`** → Unique identifier for the instance.

- For **`RecurringTask`**, the `task_id` is always `"recurring"` because there can only be one instance.  
  See code.

- For **`RecurringTaskWithPersistentSchedule`**, you can have **multiple instances** of the same task definition with different `task_id`s.
